frontend/ui/elements/page_turns.lua

```diff
@@ -1,7 +1,17 @@
 local Device = require("device")
 local Event = require("ui/event")
+local ReaderUI = require("apps/reader/readerui")
 local UIManager = require("ui/uimanager")
 local _ = require("gettext")
+local T = require("ffi/util").template
+
+local ReaderView = ReaderUI.instance.view
```

Comment:
Is this right ?!
It just reads like you're caching as a module local the instance that existed when this module was first loaded.
This might have been ok if page_turns.lua were `dofile()` (which I don't really like), but this is not what is happening - we have:
`readermenu.lua:    self.menu_items.page_turns = require("ui/elements/page_turns")`

So, you probably should fetch the current instance in each of your callbacks.

Reply:
Can I catch any errors caused with this? It is interesting to investigate them to understand the architecture.

Reply:
Sorry, not sure I understand what you are asking :/
Dunno how it was behaving, but I guess before your fix, after switching documents, you would have keep a reference to the previous ReaderUI and ReaderView instances, and you would not have acted on the old instance: may be it would have crashed, or changes to page turns and touch zones would not be reflected, or would crash when you tap or swipe.
If you're looking for other bits in the code that would have this bug, I hope there is not any :) (The one that has the most of them is frontend/ui/elements/page_overlap.lua).

Reply:
Thanks, that's what I wanted to know, how this error could show up.

Reply:
Also, to explain it a bit more:
`local ReaderUI = require("apps/reader/readerui")` stores the single module/class object.
When somewhere we do `reader = ReaderUI:new()`, we create an instance: somehow a "copy" of the class object.
When working with the instance, we use `self.something = somevalue`: this attribute is stored in the instance/copy.
But we can also store things in the single class object with `ReaderUI.something = somevalue`. All instances (the current and future) will see something we store in `ReaderUI` - but not when we store it in the instance `self`.

There are tricks in ReaderUI to stores as ReaderUI.instance (so, as an attribute of the class object) a reference to the current "instance" `reader`, so we can access it in places where we don't have access to our usual `self.ui`. This is not super proper object oriented coding, but well, we have it :)
The frontend/ui/elements bit are little modules that are require() so cached, and reused when switching documents - so they can't have access to the current `self.ui` (unless we pass it to some functions calls, if we can) - that's why they are using `ReaderUI.instance`.

Reply:
Also of interest:
https://github.com/koreader/koreader/pull/7371#discussion_r587397902
https://github.com/koreader/koreader/pull/7371#discussion_r587684916

Reply:
Adding to my encyclopedia.